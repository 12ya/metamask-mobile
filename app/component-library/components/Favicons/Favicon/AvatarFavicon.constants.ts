/* eslint-disable import/prefer-default-export */

// Third party dependencies.
import { ImageSourcePropType } from 'react-native';

export const TEST_REMOTE_IMAGE_URL = 'https://uniswap.org/favicon.ico';

export const TEST_REMOTE_IMAGE_SOURCE: ImageSourcePropType = {
  uri: TEST_REMOTE_IMAGE_URL,
};
