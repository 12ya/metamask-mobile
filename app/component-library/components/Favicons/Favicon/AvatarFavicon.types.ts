// Third party dependencies.
import { ImageSourcePropType } from 'react-native';

// External dependencies.
import { AvatarBaseProps } from '../../foundation/AvatarBase';
import { AvatarVariants } from '../../Avatar.types';

/**
 * AvatarFavicon component props.
 */
export type AvatarFaviconProps = AvatarBaseProps & {
  /**
   * Variant of Avatar
   */
  variant?: AvatarVariants.Favicon;
  /**
   * A favicon image from either a local or remote source.
   */
  imageSource: ImageSourcePropType;
};
