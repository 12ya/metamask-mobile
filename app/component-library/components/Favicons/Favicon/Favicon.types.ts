// Third party dependencies.
import { ImagePropsBase } from 'react-native';

// External dependencies.
import { CoinPatternProps } from '../../../patterns/Coins/Coin';

/**
 * Favicon sizes.
 */
export { CoinPatternSizes as FaviconSizes } from '../../../patterns/Coins/Coin/Coin.types';

/**
 * Favicon component props.
 */
export type FaviconProps = CoinPatternProps & {
  /**
   * Props for the image content.
   */
  imageProps: ImagePropsBase;
};

/**
 * Style sheet input parameters.
 */
export type FaviconStyleSheetVars = Pick<FaviconProps, 'style' | 'size'>;
