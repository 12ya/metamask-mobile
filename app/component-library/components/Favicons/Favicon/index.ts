export { default } from './Favicon';
export { FaviconSizes } from './Favicon.types';
