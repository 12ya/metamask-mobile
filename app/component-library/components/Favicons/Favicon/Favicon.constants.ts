/* eslint-disable import/prefer-default-export */
// Third party dependencies.
import { ImagePropsBase, ImageSourcePropType } from 'react-native';

// Internal dependencies.
import { FaviconProps, FaviconSizes } from './Favicon.types';

// Defaults
export const DEFAULT_FAVICON_SIZE = FaviconSizes.Md;

// Test IDs
export const FAVICON_TEST_ID = 'favicon';
export const FAVICON_IMAGE_TEST_ID = 'favicon-image';

// Sample consts
const SAMPLE_FAVICON_IMAGE_URL = 'https://uniswap.org/favicon.ico';
const SAMPLE_FAVICON_IMAGE_SOURCE: ImageSourcePropType = {
  uri: SAMPLE_FAVICON_IMAGE_URL,
};
export const SAMPLE_FAVICON_IMAGE_PROPS: ImagePropsBase = {
  source: SAMPLE_FAVICON_IMAGE_SOURCE,
};
export const SAMPLE_FAVICON_PROPS: FaviconProps = {
  imageProps: SAMPLE_FAVICON_IMAGE_PROPS,
};
