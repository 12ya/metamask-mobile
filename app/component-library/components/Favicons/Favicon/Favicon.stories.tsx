// Third party dependencies.
import React from 'react';
import { storiesOf } from '@storybook/react-native';
import { select } from '@storybook/addon-knobs';

// External dependencies.
import { storybookPropsGroupID } from '../../../../component-library/constants/storybook.constants';

// Internal dependencies.
import Favicon from './Favicon';
import {
  DEFAULT_FAVICON_SIZE,
  TEST_REMOTE_IMAGE_PROPS,
} from './Favicon.constants';
import { FaviconProps, FaviconSizes } from './Favicon.types';

export const getFaviconStoryProps = (): FaviconProps => {
  const sizeSelector = select(
    'size',
    FaviconSizes,
    DEFAULT_FAVICON_SIZE,
    storybookPropsGroupID,
  );

  return {
    size: sizeSelector,
    imageProps: TEST_REMOTE_IMAGE_PROPS,
  };
};
const FaviconStory = () => <Favicon {...getFaviconStoryProps()} />;

storiesOf('Component Library / Favicons', module).add('Favicon', FaviconStory);

export default FaviconStory;
