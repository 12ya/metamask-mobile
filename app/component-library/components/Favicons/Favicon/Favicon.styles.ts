// Third party dependencies.
import { ImageStyle, StyleSheet, ViewStyle } from 'react-native';

// External dependencies.
import { Theme } from '../../../../util/theme/models';

// Internal dependencies.
import { FaviconStyleSheetVars } from './Favicon.types';

/**
 * Style sheet function for Badge component.
 *
 * @param params Style sheet params.
 * @param params.theme App theme from ThemeContext.
 * @param params.vars Inputs that the style sheet depends on.
 * @returns StyleSheet object.
 */
const styleSheet = (params: { theme: Theme; vars: FaviconStyleSheetVars }) => {
  const { vars } = params;
  const { style, size } = vars;

  return StyleSheet.create({
    base: Object.assign({} as ViewStyle, style) as ViewStyle,
    image: Object.assign({
      width: Number(size),
      height: Number(size),
      borderRadius: Number(size) / 2,
    } as ImageStyle),
  });
};

export default styleSheet;
