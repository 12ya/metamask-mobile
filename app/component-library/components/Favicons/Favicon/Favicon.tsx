/* eslint-disable react/prop-types */

// Third party dependencies.
import React, { useCallback, useState } from 'react';
import { Image, ImageErrorEventData, NativeSyntheticEvent } from 'react-native';

// External dependencies.
import { useStyles } from '../../../hooks';
import CirclePattern from '../../../patterns/Circles/Circle/Circle';
import { IconName } from '../../../../component-library/components/Icons/Icon';
import IconContainer from '../../Icons/IconContainer';

// Internal dependencies.
import styleSheet from './Favicon.styles';
import { FaviconProps } from './Favicon.types';
import {
  DEFAULT_FAVICON_SIZE,
  FAVICON_TEST_ID,
  FAVICON_IMAGE_TEST_ID,
} from './Favicon.constants';

const Favicon = ({
  style,
  imageProps,
  size = DEFAULT_FAVICON_SIZE,
  ...props
}: FaviconProps) => {
  const [error, setError] = useState(undefined);
  const { styles } = useStyles(styleSheet, {
    style,
    size,
  });

  const onError = useCallback(
    (e: NativeSyntheticEvent<ImageErrorEventData>) =>
      setError(e.nativeEvent.error),
    [setError],
  );

  const renderError = () => (
    <IconContainer size={size} iconProps={{ name: IconName.GlobalFilled }} />
  );

  const renderImage = () => (
    <CirclePattern
      size={size}
      style={styles.base}
      testID={FAVICON_TEST_ID}
      {...props}
    >
      <Image
        resizeMode={'contain'}
        onError={onError}
        style={styles.image}
        testID={FAVICON_IMAGE_TEST_ID}
        {...imageProps}
      />
    </CirclePattern>
  );

  return <>{error ? renderError() : renderImage()}</>;
};

export default Favicon;
