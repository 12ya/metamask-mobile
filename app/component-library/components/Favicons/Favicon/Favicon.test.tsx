// Third party dependencies.
import React from 'react';
import { shallow } from 'enzyme';

// Internal dependencies.
import Favicon from './Favicon';
import { FaviconSizes } from './Favicon.types';
import {
  FAVICON_TEST_ID,
  FAVICON_IMAGE_TEST_ID,
  TEST_FAVICON_IMAGE_PROPS,
} from './Favicon.constants';

describe('Favicon - Snapshot', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <Favicon size={FaviconSizes.Xl} imageProps={TEST_FAVICON_IMAGE_PROPS} />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Favicon', () => {
  it('should render Favicon component', () => {
    const wrapper = shallow(
      <Favicon size={FaviconSizes.Xl} imageProps={TEST_FAVICON_IMAGE_PROPS} />,
    );
    const faviconComponent = wrapper.findWhere(
      (node) => node.prop('testID') === FAVICON_TEST_ID,
    );
    const faviconImageComponent = wrapper.findWhere(
      (node) => node.prop('testID') === FAVICON_IMAGE_TEST_ID,
    );
    expect(faviconComponent.exists()).toBe(true);
    expect(faviconImageComponent.exists()).toBe(true);
  });
});
