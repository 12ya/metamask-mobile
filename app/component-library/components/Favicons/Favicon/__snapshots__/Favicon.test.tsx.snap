// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favicon - Snapshot should render correctly 1`] = `
<Fragment>
  <CirclePattern
    size="48"
    style={Object {}}
    testID="favicon"
  >
    <Image
      onError={[Function]}
      resizeMode="contain"
      source={
        Object {
          "uri": "https://uniswap.org/favicon.ico",
        }
      }
      style={
        Object {
          "borderRadius": 24,
          "height": 48,
          "width": 48,
        }
      }
      testID="favicon-image"
    />
  </CirclePattern>
</Fragment>
`;
